app/ante/eth.go

```diff
@@ -17,6 +17,8 @@ import (
 	ethtypes "github.com/ethereum/go-ethereum/core/types"
 )
 
+const MinimalTxGasWanted uint64 = 21000
+
```

Comment:
Now, we are back to the issue where the proposed block includes many transactions that will never get executed. E.g. I ran the test with 50 tx of ~700k gas, all of them were included in the first block, but only 12 were executed.

I do not have a solution yet for solving both these problems, but I am thinking about:
- calculating `gasUsed` for this `gasWanted` limit in the ante handler
- looking in more detail at how blocks are processed and how `gasWanted` works in general for the cosmos sdk. I am wondering if this issue (ddos vs. including too many tx because of gasWanted) is a general cosmos sdk problem and if not, why not.

Reply:
I just saw the list of proposals here https://github.com/tharsis/ethermint/issues/989#issuecomment-1067514948 and the long term solution sounds right
> We can run the tx in the PrepareProposal phase, and filter the preliminary transaction list according to the gas used.

This temporary fix is ok from my side.

Reply:
> Now, we are back to the issue where the proposed block includes many transactions that will never get executed. E.g. I ran the test with 50 tx of ~700k gas, all of them were included in the first block, but only 12 were executed.
> 
> 
> 
> I do not have a solution yet for solving both these problems, but I am thinking about:
> 
> - calculating `gasUsed` for this `gasWanted` limit in the ante handler
> 
> - looking in more detail at how blocks are processed and how `gasWanted` works in general for the cosmos sdk. I am wondering if this issue (ddos vs. including too many tx because of gasWanted) is a general cosmos sdk problem and if not, why not.

Cosmos-sdk don't refund unused gases, so it's not an issue there.

Reply:
Although this is treated as a temporary fix, I want to propose if we can make this slightly smarter. For example

1. Can we peek into the `data` field of the EVM Tx, if it is a simple transfer with no data, `gasWanted` is 21000
2. Otherwise, calculate the `gasWanted` as `gasLimit / n` where n can be further decided.

An attacker could still specify a very large gas limit with this, so this approach would potentially require to charge the sender the `MAX(gasWanted, gasLimit/n)` to create a base line of defence ... something borrowed from #989.

However, this requires a change of the gas refund approach, which is a bigger decision to make than the tx inclusion logic. Charging the user the `gasPrice*gasLimit` is common on Cosmos world but probably not on ETH world.

Reply:
how about still use the `gasLimit` but put a maximum limitation so that there is still a "minimum" of transaction that can be accepted?

Someone wouldnt be able to DOS without spending fair amount of gas in that case?

Reply:
put maximum limitation to `gasLimit` sounds good.

Reply:
Isn't this potentially breaking? In a theoretical world there could be a codepath that requires more than 500k gas to execute. If a chain upgrades to this version of ethermint it could fully brick someones smart contract. 

Temporarily I guess it's *fine* though

Reply:
> Isn't this potentially breaking? In a theoretical world there could be a codepath that requires more than 500k gas to execute. If a chain upgrades to this version of ethermint it could fully brick someones smart contract.
> 
> Temporarily I guess it's _fine_ though

it doesn't affect the tx execution, because the gas meter is infinite, so only affects the tx inclusion.

Reply:
ahh yes true, the evm uses infinite, thanks yihuang